src/Storages/MergeTree/MergeTreeData.h

```diff
@@ -914,7 +914,7 @@ class MergeTreeData : public IStorage, public WithMutableContext
     /// section from config.xml.
     CompressionCodecPtr getCompressionCodecForPart(size_t part_size_compressed, const IMergeTreeDataPart::TTLInfos & ttl_infos, time_t current_time) const;
 
-    std::lock_guard<std::mutex> getQueryIdSetLock() const { return std::lock_guard<std::mutex>(query_id_set_mutex); }
+    std::mutex & getQueryIdSetLock() const { return query_id_set_mutex; }
```

Comment:
What is the purpose of this change? We have a lot of methods that return different kinds of "lock holders" and it's quite convenient and it usually looks better than returning reference to a mutex and locking it from outside. And in more complex cases (like `DatabaseCatalog::getDDLGuard` or `IStorage::lockForShare`) code from outside should not know how to lock and unlock "mutex", it should only get lock holder object and call its dtor

Reply:
Short story: TSA is confused. Will revert to the previous locking and disable TSA.

Long story: It's a bit tricky, but TSA gets confused. It does not understand that the following code is protected if we call `getQueryIdSetLock()` + hold the return value (= a `std::lock_guard`).

My first "fix" was to disable TSA locally at the caller via pragmas (disable `-Wthread-safety-analysis`) , but that wasn't very pretty.

This is the second try: expose the mutex and let the caller do the locking. Clearly, this is more dangerous than before. In this case, the safety net is that the `NoLock*()` methods (= methods that require holding the lock) take a `std::lock_guard<std::mutex> &` dummy parameter - and they do that already without this commit. TSA should in theory be able to enforce the same using the `RETURN_CAPABILITY` annotation ([0]) but I couldn't really get that to work.

I agree that returning a `std::lock_guard` is more convenient + safer than exposing the mutex. Will therefore roll back to the previous locking. I think that we can nevertheless remove the existing "safety net" (dummy `std::lock_guard` params in `NoLock*()` methods). Why? Because if we annotate the shared data in `MergeTreeData` as protected by mutex `query_id_set_mutex`, then TSA will complain about every unsynchronized access to it. In this case, the programmer will get compile errors which can only be resolved by disabling TSA using annotation `TSA_NO_THREAD_SAFETY_ANALYSIS`, similar to the first try. It is not very beautiful  optimal solution but at least it is impossible to miss locking problems.

(The fourth approach would be to not annotate the members in `MergeTreeData` at all. But as this commit should demonstrate the features of TSA, I would rather like to not do that. )

[0] https://clang.llvm.org/docs/ThreadSafetyAnalysis.html